src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -89,6 +90,15 @@ static bool exclusive = false;
 #endif
 #define htonll(a) ntohll(a)
 
+static int do_unmap();
+
+static void handle_signal(int signum)
+{
+  assert(signum == SIGINT || signum == SIGTERM);
+  derr << "*** Got signal " << sig_str(signum) << " ***" << dendl;
+  do_unmap();
```

Comment:
After you move register/unregister closer to `ioctl(nbd, NBD_DO_IT)` a race seams like not possible (previously a scenario was possible, when the main thread closes nbd, a new process reuses this device, and then a signal handler fires sending the disconnect to a wrong device).

So the main point now is just to make it lightweight: no need to open/close a device if you can just use a descriptor. I wont insist though if you like your version more.